types/block.go

```diff
@@ -259,6 +264,128 @@ func mustPush(rowTree *nmt.NamespacedMerkleTree, id namespace.ID, data []byte) {
 	}
 }
 
+func (b *Block) PutBlock(ctx context.Context, api ipfsapi.CoreAPI) error {
+	if api == nil {
+		return errors.New("no ipfs api object provided")
+	}
+
+	// recomputing the erasured data
+	namespacedShares := b.Data.computeShares()
+	shares := namespacedShares.RawShares()
+	if len(shares) == 0 {
+		// no shares -> no row/colum roots -> hash(empty)
+		b.DataHash = b.DataAvailabilityHeader.Hash()
+		return nil
+	}
+
+	extendedDataSquare, err := rsmt2d.ComputeExtendedDataSquare(shares, rsmt2d.RSGF8, rsmt2d.NewDefaultTree)
+	if err != nil {
+		panic(fmt.Sprintf("unexpected error: %v", err))
+	}
+
+	squareWidth := extendedDataSquare.Width()
+	originalDataWidth := squareWidth / 2
+
+	// add namespaces to erasured shares and chunk into tree sized portions
+	leaves := make([][][]byte, 2*squareWidth)
+	for outerIdx := uint(0); outerIdx < squareWidth; outerIdx++ {
+		rowLeaves := make([][]byte, squareWidth)
+		colLeaves := make([][]byte, squareWidth)
+		for innerIdx := uint(0); innerIdx < squareWidth; innerIdx++ {
+			if outerIdx < originalDataWidth && innerIdx < originalDataWidth {
+				rowShare := namespacedShares[outerIdx*originalDataWidth+innerIdx]
+				colShare := namespacedShares[innerIdx*originalDataWidth+outerIdx]
+				rowLeaves[innerIdx] = append(rowShare.NamespaceID(), rowShare.Data()...)
+				colLeaves[innerIdx] = append(colShare.NamespaceID(), colShare.Data()...)
+			} else {
+				rowData := extendedDataSquare.Row(outerIdx)
+				colData := extendedDataSquare.Column(outerIdx)
+				parityCellFromRow := rowData[innerIdx]
+				parityCellFromCol := colData[innerIdx]
+				rowLeaves[innerIdx] = append(copyOfParityNamespaceID(), parityCellFromRow...)
+				colLeaves[innerIdx] = append(copyOfParityNamespaceID(), parityCellFromCol...)
+			}
+		}
+		leaves[outerIdx] = rowLeaves
+		leaves[2*outerIdx] = colLeaves
+	}
```

Comment:
Might be orthogonal to this PR but we should either refactor this s.t. this is not duplicated here, or, use only the rsmt2d lib for getting the row/column roots (the latter is preferable if it works). 

Or, we could also just use the NMT + a "node collector" function here directly and skip using `DataSquareRowOrColumnRawInputParser`). 

https://github.com/lazyledger/lazyledger-core/blob/899f5b23d738e6d0118fddfbd3c0410d2f562834/types/block.go#L213-L244

Reply:
That is a good idea! I ended up refactoring out the redundant code. https://github.com/lazyledger/lazyledger-core/pull/178/commits/5bfb03f867f8ff20d3c00548bfcd4edb5a9c55e1 I don't think we can use the rsmt2d libs, because we have to add the namespaces back to the data before generating the roots, but I could be missing something. We could definitely use the node collector in this package instead of `DataSquarerowOrColumnRawInputParser`, but I didn't for the time being in order not to export `nmtNode` from the plugin lib.

Reply:
> I don't think we can use the rsmt2d libs, because we have to add the namespaces back to the data before generating the roots,

Alright! Yeah, I knew there was a catch but I couldn't remember what it was. We could add this kind of processing to the rsmt2d lib though: https://github.com/lazyledger/rsmt2d/issues/12#issuecomment-763576720

But yeah, this is also completely orthogonal to this PR. Thanks for the refactoring! Looks good so far.

> We could definitely use the node collector in this package instead of DataSquarerowOrColumnRawInputParser, but I didn't for the time being in order not to export nmtNode from the plugin lib.

Couldn't we simply export the NodeCollector and keep the nmtNode private?

Reply:
Good idea. Is this what you had in mind? https://github.com/lazyledger/lazyledger-core/pull/178/commits/47b4125538e38f7781ae4de06cf620706aa55c2f

Reply:
Yes, that looks good!